### PR TITLE
Songs rooms

### PR DESCRIPTION
**Pull request by tshuck**
@sisk @DLavin23 

Cleans up songs, rooms relationships:
  - Remove direct relationship between songs, room
  - Allows songs to be find-or-created by youtube id
  - Adds a RoomQueue model to associate a song, room, and user.

tumblr_n17r5jTMa11qmloy1o2_400

### CODE REVIEW

**Comment by sisk** [comment]
Reviewing. My WIP is [here](https://github.com/tshuck/musicbox-api/commit/18130a950ba64174c51aa723293681f454b5a3f3#diff-90f827681ccbedf6cbfabf956112dc89R5) to show how the Youtube gem works.

File: app/controllers/songs_controller.rb [hunk 1]
```diff
@@ -12,12 +12,20 @@ class SongsController < ApplicationController
...
     else
       render_errors_for(song)
     end
   end
+
+  private
+
+  def persisted_song
+    return @_persisted_song if defined? @_persisted_song
+    @_persisted_song = Song.find_by(youtube_id: attrs[:youtube_id])
```
Comment (sisk) [inline app/controllers/songs_controller.rb:29]: Do we always assume `attrs[:youtube_id]` is provided? If not, it's nil, and then the persisted song that gets memoized here is, I think, the first one where `youtube_id` is nil in that case. That seems logically wrong.

**Review by sisk — COMMENTED** [review]
Looks good in general. I like a "queue" join resource here...cleans up our need to have some overlap between models and still have a central ordering source. I think there may be a logical flaw in the song create action. Please review to make sure we're not memoizing songs with null youtube_id by default.

**Comment by tshuck** [comment]
@sisk Good catch.  Added a validation to require `youtube_id`.

**Pull request merged**